functions/db/deliveryRequests/onCreate.f.js

```diff
@@ -111,23 +111,64 @@ function getRASnapPromise(accountsRef, raId) {
-    var requestEnd = moment(request.endTime, enums.DateTimeFormat.TIME);
+    // ------- helpers --------
+    var loggingF = 'ddd YYYY-MM-DD HH:mm:ssZ';  // 'Mon 2018-04-09 21:00:00-07:00'
+    var timeF = 'HH:mm';
+    var dateF = 'YYYY-MM-DD';
+    var dateTimeF = dateF + timeF;
+
+    function constructRaTimeInReq(raTimestamp, reqInRA) {
+        // get RA's weekday and time in RA's own timezone first
+        var raTimeOriginal = moment.tz(raTimestamp, ra.timezone);
+        // make ra time's date the same as request's (in RA timezone)
+        var raTime = moment.tz(
+            reqInRA.format(dateF) + raTimeOriginal.format(timeF), dateTimeF, ra.timezone);
+        raTime.tz(request.timezone);
+        return raTime;
+    }
+    // ------- end helpers --------
+
+    // get request start/end time in request's original timezone
+    // Eg: Thu 3/1/2018 21:00 (REQ) - Thu 5/3/2018 23:00 (REQ)
+    var reqStart = moment.tz(request.startTimestamp, request.timezone);
+    var reqEndOriginal = moment.tz(request.endTimestamp, request.timezone);
+    console.info('[isAvailable] Req(' + request.timezone + '): ' +
+        reqStart.format(loggingF) + ' - ' + reqEndOriginal.format(loggingF));
+
+    // get the weekday of request in RA's timezone, so that we'd grab the
+    // availability for the right day
+    // Eg: RA(ra_timezone) = REQ(req_timezone) + 5hrs
+    //     "Thu 3/1/2018 21:00 (REQ)" = "Fri 3/2/2018 02:00 (RA)"
+    var reqInRA = moment.tz(request.startTimestamp, ra.timezone);
+    var raDay = ra.availabilities[reqInRA.day()];
+    console.info('[isAvailable] Req in RA(' + ra.timezone + '): ' + reqInRA.format(loggingF));
 
     // raDay could be null if no available slots for that day
-    var raDay = ra.availabilities[requestDay];
-
     if (raDay) {
-        var raStart = moment(raDay.startTime, enums.DateTimeFormat.TIME);
-        var raEnd = moment(raDay.endTime, enums.DateTimeFormat.TIME);
-
-        if (requestStart.isSameOrAfter(raStart) &&
-            requestEnd.isSameOrBefore(raEnd)) {
+        // Compare everything in request's timezone.
+        // Eg: RA's availability: "Fri 08:00 to 14:00 (RA)" 
+        //                     => "Fri 03:00 to 09:00 (REQ)"
+        // At this point, we want to compare:
+        //    Request: Thu 3/1/2018 21:00 - 23:00 (REQ)
+        //    RA:      Fri 3/2/2018 03:00 - 09:00 (REQ)
+        var reqEnd = moment.tz(
+            reqStart.format(dateF) + reqEndOriginal.format(timeF), dateTimeF, request.timezone);
+        var raStart = constructRaTimeInReq(raDay.startTimestamp, reqInRA);
+        var raEnd = constructRaTimeInReq(raDay.endTimestamp, reqInRA);
+
+        console.info(
+            '[isAvailable] Req: ' + reqStart.format(loggingF) + ' - ' + reqEnd.format(loggingF));
+        console.info(
+            '[isAvailable] RA: ' + raStart.format(loggingF) + ' - ' + raEnd.format(loggingF));
+
+        if (reqStart.isSameOrAfter(raStart) &&
```

Comment:
Just out of curiosity: do we need to take into account the __delivery time__, which might be 3 hours or something, when we are determining the availability?

Reply:
oh man, you mean like once they pick up the delivery, they have 3 hours to deliver it thing? ummmmmmmm I'm gonna say no for now 😂 I was under the impression that the `start time` and `end time` is the entire duration of the delivery (picking up and delivered have to all happen within that window), but now i'm not so sure..
I think we should stop dwelling on this availability for now lol this is a basic working version of the functionality and I'll add a TODO to clarify with Madison later. What do you think?

Reply:
I think it's worth considering but fine to leave out. Essentially, the goal is to notify RAs who are _potentially_ available for pickup. They can decline for any number of reasons, and I think not having time to deliver it is a valid reason to decline. For now we can make it the RAs responsibility to take delivery time into account and have the human on the RAs side decide whether to accept the request or not.